mayastor/src/subsys/mbus/registration.rs

```diff
@@ -0,0 +1,201 @@
+//! Registration subsystem connecting mayastor to control plane (moac).
+//! A registration message is used to let the control plane know about
+//! a mayastor instance.
+//! A deregistration message is used let the control plane know that
+//! a mayastor instance is going down.
+//!
+//! The registration messages are currently sent on an `HB_INTERVAL`
+//! containing the node name and the grpc endpoint.
+
+use futures::{select, FutureExt, StreamExt};
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::{env, time::Duration};
+
+use crate::subsys::MessageBus;
+
+/// Mayastor sends registration messages in this interval (kind of heart-beat)
+const HB_INTERVAL: Duration = Duration::from_secs(2);
+
+/// Errors for pool operations.
+///
+/// Note: The types here that would be normally used as source for snafu errors
+/// do not implement Error trait required by Snafu. So they are renamed to
+/// "cause" attribute and we use .map_err() instead of .context() when creating
+/// them.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to connect to the MessageBus server {}: {:?}",
+        server,
+        cause
+    ))]
+    ConnectFailed {
+        cause: std::io::Error,
+        server: String,
+    },
+    #[snafu(display(
+        "Cannot issue requests if message bus hasn't been started"
+    ))]
+    NotStarted {},
+    #[snafu(display("Failed to queue register request: {:?}", cause))]
+    QueueRegister { cause: std::io::Error },
+    #[snafu(display("Failed to queue deregister request: {:?}", cause))]
+    QueueDeregister { cause: std::io::Error },
+}
+
+/// Register message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct RegisterArgs {
+    id: String,
+    #[serde(rename = "grpcEndpoint")]
+    grpc_endpoint: String,
+}
+
+/// Deregister message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct DeregisterArgs {
+    id: String,
+}
+
+#[derive(Clone)]
+struct Configuration {
+    /// Name of the node that mayastor is running on
+    node: String,
+    /// gRPC endpoint of the server provided by mayastor
+    grpc_endpoint: String,
+    /// heartbeat interval (how often the register message is sent)
+    hb_interval: Duration,
+}
+
+#[derive(Clone)]
+pub(super) struct Registration {
+    /// Configuration of the registration
+    config: Configuration,
+    /// Receive channel for messages and termination
+    rcv_chan: smol::channel::Receiver<()>,
+    /// Termination channel
+    fini_chan: smol::channel::Sender<()>,
+}
+
+static MESSAGE_BUS_REG: OnceCell<Registration> = OnceCell::new();
+impl Registration {
+    pub(super) fn init(node: &str, grpc_endpoint: &str) {
+        MESSAGE_BUS_REG.get_or_init(|| Registration::new(node, grpc_endpoint));
+
+        // spawn a runner thread responsible for registering and
+        // deregistering the mayastor instance on shutdown
+        std::thread::spawn(|| {
```

Comment:
hmpf. Why do we need a thread for this? and if we do, we should set its mask such that it does not run on the coremask.

Reply:
The registration process keeps running waking up on an interval to do work so we can't run on the "subsystem init context". 
Good catch, guess we could use `spawn_unaffinitized`.

Reply:
Yeah but as we use futures, can't we use a timer to ping pong here for us?

Reply:
What do you mean by ping pong?